tests/units/units.cpp

```diff
@@ -1060,19 +1060,132 @@ TEST(Units, scalingFactorWithTwoEmptyUnits)
+//     libcellml::UnitsPtr u2 = libcellml::Units::create();
+//     u2->setName("u2");
+//     u2->addUnit("metre", 0, 1.0, 1.0);
+//     u2->addUnit("metre", 0, 1.0, 1.0);
+//     u2->addUnit("metre", 0, 1.0, 1.0);
+
+//     EXPECT_EQ(1000.0, libcellml::Units::scalingFactor(u2, u1));
+//     EXPECT_EQ(0.001, libcellml::Units::scalingFactor(u1, u2));
+// }
+
+TEST(Units, scalingFactorBetweenDissimilarUnits)
+{
+    libcellml::UnitsPtr u1 = libcellml::Units::create();
+    u1->setName("u1");
+    u1->addUnit("metre", 0, 1.0, 1000.0);
+    libcellml::UnitsPtr u2 = libcellml::Units::create();
+    u2->setName("u2");
+    u2->addUnit("metre", 0, 1.0, 1.0);
+    u2->addUnit("metre", 0, 1.0, 1.0);
+    u2->addUnit("metre", 0, 1.0, 1.0);
+
+    EXPECT_EQ(0.0, libcellml::Units::scalingFactor(u2, u1));
+}
+
+TEST(Units, compareIncompatibleMultiplierSimple)
 {
-    // u1 = 1000*u2
     libcellml::UnitsPtr u1 = libcellml::Units::create();
     u1->setName("u1");
     u1->addUnit("metre", 0, 1.0, 1.0);
     libcellml::UnitsPtr u2 = libcellml::Units::create();
     u2->setName("u2");
     u2->addUnit("second", 0, 1.0, 1.0);
 
-    // Incompatible units but return a scaling factor of 1.0.
+    // Incompatible units return a scaling factor of 0.0.
+    EXPECT_EQ(0.0, libcellml::Units::scalingFactor(u1, u2));
+    EXPECT_EQ(0.0, libcellml::Units::scalingFactor(u2, u1));
+}
+
+TEST(Units, dimensionlessScalingFactor)
+{
+    auto m = libcellml::Model::create("model");
+    auto u = libcellml::Units::create("u");
+
+    auto u1 = libcellml::Units::create("u1");
+    u1->addUnit("u", "milli", 2.0, 1000.0);
+    u1->addUnit("dimensionless");
+    auto u2 = libcellml::Units::create("u2");
+    u2->addUnit("u", "kilo", 2.0, 0.001);
+    u2->addUnit("dimensionless");
+
+    auto u3 = libcellml::Units::create("u3");
+    u3->addUnit("u", "milli", 2.0, 1000.0);
+    auto u4 = libcellml::Units::create("u4");
+    u4->addUnit("u", "kilo", 2.0, 0.001);
+
+    auto u5 = libcellml::Units::create("u5");
+    u5->addUnit("u", "milli", 2.0, 1000.0);
+    u5->addUnit("dimensionless", 2.0);
+    auto u6 = libcellml::Units::create("u6");
+    u6->addUnit("u", "kilo", 2.0, 0.001);
+    u6->addUnit("dimensionless", 2.0);
+
+    m->addUnits(u);
+    m->addUnits(u1);
+    m->addUnits(u2);
+    m->addUnits(u3);
+    m->addUnits(u4);
+    m->addUnits(u5);
+    m->addUnits(u6);
+
     EXPECT_EQ(1.0, libcellml::Units::scalingFactor(u1, u2));
     EXPECT_EQ(1.0, libcellml::Units::scalingFactor(u2, u1));
+    EXPECT_EQ(1.0, libcellml::Units::scalingFactor(u3, u3));
+    EXPECT_EQ(1.0, libcellml::Units::scalingFactor(u4, u4));
+    EXPECT_EQ(1.0, libcellml::Units::scalingFactor(u1, u3));
+    EXPECT_EQ(1.0, libcellml::Units::scalingFactor(u2, u4));
+    EXPECT_EQ(1.0, libcellml::Units::scalingFactor(u1, u4));
+    EXPECT_EQ(1.0, libcellml::Units::scalingFactor(u2, u3));
+    EXPECT_EQ(1.0, libcellml::Units::scalingFactor(u5, u6));
+    EXPECT_EQ(1.0, libcellml::Units::scalingFactor(u1, u5));
```

Comment:
The tests between `u2` and `u4`, `u2` and `u3` are not needed. Also, we might want to add a test between `u6` and `u5`. In the end, I would simply have:

```
EXPECT_EQ(1.0, libcellml::Units::scalingFactor(u1, u2));
EXPECT_EQ(1.0, libcellml::Units::scalingFactor(u2, u1));
EXPECT_EQ(1.0, libcellml::Units::scalingFactor(u3, u4));
EXPECT_EQ(1.0, libcellml::Units::scalingFactor(u4, u3));
EXPECT_EQ(1.0, libcellml::Units::scalingFactor(u5, u6));
EXPECT_EQ(1.0, libcellml::Units::scalingFactor(u6, u5));
EXPECT_EQ(1.0, libcellml::Units::scalingFactor(u1, u3));
EXPECT_EQ(1.0, libcellml::Units::scalingFactor(u1, u5));
```

Reply:
u2, u4 -> comparison needed because previously a mismatch of "dimensionless" caused this test to fail, same with u2, u3.  I'll add the test from u6 to u5 as you suggest.